design_documents/parametrizing_tests_with_fixed_values.md

```diff
@@ -0,0 +1,192 @@
+This way, if the user provides invalid value for the type defined, the compiler will throw a relevant error.
+
+### Parametrization With Fixtures
+
+```cairo
+struct MyStruct {
+    a: felt252, 
+    b: Array<felt252>
+}
+
+fn my_fixture() -> MyStruct {
+    MyStruct { a: 1, b: array![2, 3] }
+}
+
+#[parametrize]
+#[case(1, my_fixture())]
+#[case(3, my_fixture())]
+#[test]
+fn my_test(a: felt252, b: MyStruct) {
+    // case a = 1, b = MyStruct { a: 1, b: [2, 3] }
+    // case a = 3, b = MyStruct { a: 1, b: [2, 3] }
+    // ...
+}
+```
+
+For fixtures, we would generate code in the exactly same manner as for simple arguments.
+
+```cairo
+// ...
+#[test]
+fn my_test_1() {
+    let a = 1;
+    let b = my_fixture();
+    // ...
+}
+```
+
+### Parametrized Fixtures
+
+```cairo
+struct MyStruct {
+    a: felt252, 
+    b: Array<felt252>
+}
+
+fn my_fixture(a: felt252, b: felt252, c: felt252) -> MyStruct {
+    MyStruct { a: a, b: array![b, c] }
+}
+
+#[parametrize]
+#[case(1, my_fixture(2, 3, 4))]
+#[case(3, my_fixture(5, 6, 7))]
+#[test]
+fn my_test(a: felt252, b: MyStruct) {
+    // case a = 1, b = MyStruct { a: 2, b: [3, 4] }
+    // case a = 3, b = MyStruct { a: 5, b: [6, 7] }
+    // ...
+}
+```
+
+This could be handled in a very similar manner as the [non-parametrized fixture case](#parametrization-with-fixtures):
+
+```cairo
+// ...
+#[test]
+fn my_test_generated(a: felt252) {
+    let a = 1;
+    let b = my_fixture(2, 3, 4);
+    // ...
+}
+```
+
+### Parametrizing With Complex Types
+
+While it is technically possible to use complex types in attributes instead of defining fixtures, handling of them in
+code would be quite problematic based on my limited research.
+
+At least initially, we do not want to support syntax with complex types. For example, this would not be allowed:
+
+```cairo
+#[parametrize]
+#[case(MyStruct { a: 1, b: array![2, 3] })]
+#[test]
+fn my_test(a: MyStruct) {
+    // ...
+}
+```
+
+### Problems With Code Generation
+
+The code generation solution has some problems that will need to be addressed before it is implemented.
+See [section below](#possible-problems-with-code-generation) for details.
+
+## User Experience
+
+Unlike fuzz-tests, we should run and indicate the result of all parametrized cases.
+If some test cases fail, others should still be executed and their results should be displayed to the user.
+
+An example output could look similarly to this:
+
+```shell
+$ snforge test
+[PASS] tests::parametrized(a = 1, b = 2, c = my_fixture(1, 2))
+[PASS] tests::parametrized(a = 3, b = 5, c = my_fixture(4, 5))
+[FAIL] tests::parametrized(a = 4, b = 5, c = my_fixture(3, 2))
+
+Failure data:
+    original value: [344693033283], converted to a string: [PANIC]
+    
+[PASS] tests::parametrized(a = 5, b = 7, c = my_fixture(7, 8))
+# ...
+```
+
+## Required Changes to Test Collector
+
+To support parametric test, we need to introduce some changes to the test collector in Scarb.
```

Comment:
What changes?

Reply:
There are subheaders below this outlining the changes

Reply:
Sorry, hard to read in without markdown formatting 😄